crates/sparrow-backend/src/exprs/expr_pattern.rs

```diff
@@ -0,0 +1,166 @@
+use std::str::FromStr;
+
+use smallvec::{smallvec, SmallVec};
+use sparrow_arrow::scalar_value::ScalarValue;
+
+use crate::exprs::expr_lang::ExprLang;
+use crate::exprs::ExprVec;
+use crate::Error;
+
+/// A representation of an expression with "holes" that may be instantiated.
+#[derive(Debug, Default)]
+pub(crate) struct ExprPattern {
+    pub(super) expr: egg::PatternAst<ExprLang>,
+}
+
+#[static_init::dynamic]
+pub(crate) static INPUT_VAR: egg::Var = egg::Var::from_str("?input").unwrap();
+
+impl ExprPattern {
+    /// Create a new `ExprPattern` which is an identity expression referencing the input.
+    pub fn new_input() -> error_stack::Result<Self, Error> {
+        let mut exprs = ExprPattern::default();
+        exprs.add_var(*INPUT_VAR)?;
+        Ok(exprs)
+    }
+
+    /// Create an expr pattern containing the given arguments.
+    ///
+    /// Returns the resulting pattern as well as the IDs of each of the arguments.
+    pub fn new_instruction(
+        name: &'static str,
+        literal_args: smallvec::SmallVec<[ScalarValue; 2]>,
+        args: Vec<ExprPattern>,
+        data_type: arrow_schema::DataType,
+    ) -> error_stack::Result<ExprPattern, Error> {
+        let mut graph = egg::EGraph::<egg::ENodeOrVar<ExprLang>, ()>::default();
+        let mut arg_ids = SmallVec::with_capacity(args.len());
+        for arg in args {
+            let id = graph.add_expr(&arg.expr);
+            arg_ids.push(id);
+        }
+
+        // We can only extract a single expression, so we create one.
+        // This is why we need to know the instruction to create, rather than
+        // just returning the resulting `egg::Id` for each argument.
+        let output = graph.add(egg::ENodeOrVar::ENode(ExprLang {
+            name,
+            literal_args,
+            args: arg_ids,
+            result_type: data_type,
+        }));
+
+        let cost_function = egg::AstSize;
+        let extractor = egg::Extractor::new(&graph, cost_function);
```

Comment:
Kind of? The `ExprLang` is passed to `egg::EGraph<ExprLang>` and `egg::RecExpr<ExprLang>`, and tells it what nodes to use. `matches` is used to identify equivalent nodes.

Eg., if we add `add(a, b)` twice we get the same node back.

This will be most useful when we add "simplifications" so we can say "`add(a, b)` is equivalent to `add(b, a)`" and these are recorded as equivalent. Then, `foo(add(a, b))` will automatically be equivalent to `foo(add(b, a))` because it is a matching operator (`foo`) on equivalent arguments.